### ISSUE TITLE
heap-use-after-free in storage_proxy::remote::handle_write during shutdown hit by lwt test_bank_with_nemesis dtest

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylladb#16771]
Seen in https://jenkins.scylladb.com/job/scylla-master/job/dtest-debug/276/artifact/logs-heavy.debug.002/1705215764355_lwt_banking_load_test.py%3A%3ATestLWTBankingLoad%3A%3Atest_bank_with_nemesis/node3.log
during shutdown
```
Scylla version 5.5.0~dev-0.20240112.4e18f8b45381 with build-id d5c28c468be96f95c71317049ad35cd1814a67dc starting ...

WARN  2024-01-14 07:02:37,290 [shard 0:stmt] storage_proxy - Failed to apply mutation from 127.0.26.4#0: exceptions::mutation_write_timeout_exception (Operation timed out for system.paxos - received only 0 responses from 1 CL=ONE.)
=================================================================
==1475==ERROR: AddressSanitizer: heap-use-after-free on address 0x608000c293a8 at pc 0x000010f4c4d1 bp 0x7ffc5cf842d0 sp 0x7ffc5cf842c8
READ of size 8 at 0x608000c293a8 thread T0
*** stack smashing detected ***: terminated
    #0 0x10f4c4d0  (/jenkins/workspace/scylla-master/dtest-debug/scylla/.ccm/scylla-repository/4e18f8b45381c6b23cf271cc686194d87bf212a1/libexec/scylla+0x10f4c4d0) (BuildId: d5c28c468be96f95c71317049ad35cd1814a67dc)
    #1 0x10da4d45  (/jenkins/workspace/scylla-master/dtest-debug/scylla/.ccm/scylla-repository/4e18f8b45381c6b23cf271cc686194d87bf212a1/libexec/scylla+0x10da4d45) (BuildId: d5c28c468be96f95c71317049ad35cd1814a67dc)
    #2 0x10f5171d  (/jenkins/workspace/scylla-master/dtest-debug/scylla/.ccm/scylla-repository/4e18f8b45381c6b23cf271cc686194d87bf212a1/libexec/scylla+0x10f5171d) (BuildId: d5c28c468be96f95c71317049ad35cd1814a67dc)
    #3 0x7f73352e90e8  (/jenkins/workspace/scylla-master/dtest-debug/scylla/.ccm/scylla-repository/4e18f8b45381c6b23cf271cc686194d87bf212a1/libreloc/libseastar.so+0x46e90e8) (BuildId: 95698310d1473fda2254677214f6407401d5374d)
    #4 0x7f73352f1558  (/jenkins/workspace/scylla-master/dtest-debug/scylla/.ccm/scylla-repository/4e18f8b45381c6b23cf271cc686194d87bf212a1/libreloc/libseastar.so+0x46f1558) (BuildId: 95698310d1473fda2254677214f6407401d5374d)
    #5 0x7f73352f55f0  (/jenkins/workspace/scylla-master/dtest-debug/scylla/.ccm/scylla-repository/4e18f8b45381c6b23cf271cc686194d87bf212a1/libreloc/libseastar.so+0x46f55f0) (BuildId: 95698310d1473fda2254677214f6407401d5374d)
    #6 0x7f73352f3273  (/jenkins/workspace/scylla-master/dtest-debug/scylla/.ccm/scylla-repository/4e18f8b45381c6b23cf271cc686194d87bf212a1/libreloc/libseastar.so+0x46f3273) (BuildId: 95698310d1473fda2254677214f6407401d5374d)
    #7 0x7f7334e2caba  (/jenkins/workspace/scylla-master/dtest-debug/scylla/.ccm/scylla-repository/4e18f8b45381c6b23cf271cc686194d87bf212a1/libreloc/libseastar.so+0x422caba) (BuildId: 95698310d1473fda2254677214f6407401d5374d)
    #8 0x7f7334e2a05a  (/jenkins/workspace/scylla-master/dtest-debug/scylla/.ccm/scylla-repository/4e18f8b45381c6b23cf271cc686194d87bf212a1/libreloc/libseastar.so+0x422a05a) (BuildId: 95698310d1473fda2254677214f6407401d5374d)
    #9 0xd270d49  (/jenkins/workspace/scylla-master/dtest-debug/scylla/.ccm/scylla-repository/4e18f8b45381c6b23cf271cc686194d87bf212a1/libexec/scylla+0xd270d49) (BuildId: d5c28c468be96f95c71317049ad35cd1814a67dc)
    #10 0xd26dc2f  (/jenkins/workspace/scylla-master/dtest-debug/scylla/.ccm/scylla-repository/4e18f8b45381c6b23cf271cc686194d87bf212a1/libexec/scylla+0xd26dc2f) (BuildId: d5c28c468be96f95c71317049ad35cd1814a67dc)
    #11 0x7f732ea49b89  (/jenkins/workspace/scylla-master/dtest-debug/scylla/.ccm/scylla-repository/4e18f8b45381c6b23cf271cc686194d87bf212a1/libreloc/libc.so.6+0x27b89) (BuildId: 7026fe8c129a523e07856d7c96306663ceab6e24)
    #12 0x7f732ea49c4a  (/jenkins/workspace/scylla-master/dtest-debug/scylla/.ccm/scylla-repository/4e18f8b45381c6b23cf271cc686194d87bf212a1/libreloc/libc.so.6+0x27c4a) (BuildId: 7026fe8c129a523e07856d7c96306663ceab6e24)
    #13 0xd191c64  (/jenkins/workspace/scylla-master/dtest-debug/scylla/.ccm/scylla-repository/4e18f8b45381c6b23cf271cc686194d87bf212a1/libexec/scylla+0xd191c64) (BuildId: d5c28c468be96f95c71317049ad35cd1814a67dc)
```

Decoded:
```
service::storage_proxy::remote::send_mutation_failed(netw::msg_addr, tracing::trace_state_ptr, unsigned int, unsigned long, unsigned long, db::view::update_backlog, replica::exception_variant) at ./service/storage_proxy.cc:286
seastar::future<seastar::rpc::no_wait_type> service::storage_proxy::remote::handle_write<utils::tagged_uuid<table_schema_version_tag>, service::paxos::proposal, service::storage_proxy::remote::handle_paxos_learn(seastar::rpc::client_info const&, seastar::rpc::opt_time_point, service::paxos::proposal, utils::small_vector<gms::inet_address, 3ul>, gms::inet_address, unsigned int, unsigned long, std::optional<tracing::trace_info>)::{lambda(seastar::shared_ptr<service::storage_proxy>&, tracing::trace_state_ptr, seastar::lw_shared_ptr<schema const>, service::paxos::proposal const&, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >, service::fencing_token)#1}, service::storage_proxy::remote::handle_paxos_learn(seastar::rpc::client_info const&, seastar::rpc::opt_time_point, service::paxos::proposal, utils::small_vector<gms::inet_address, 3ul>, gms::inet_address, unsigned int, unsigned long, std::optional<tracing::trace_info>)::{lambda(seastar::shared_ptr<service::storage_proxy>&, netw::msg_addr, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >, service::paxos::proposal const&, gms::inet_address, unsigned int, unsigned long, std::optional<tracing::trace_info> const&, service::fencing_token)#1}>(netw, seastar::rpc::opt_time_point, utils::tagged_uuid<table_schema_version_tag>, service::paxos::proposal, utils::small_vector<gms::inet_address, 3ul> const&, gms::inet_address, unsigned int, unsigned long, std::optional<tracing::trace_info> const, service::fencing_token, service::storage_proxy::remote::handle_paxos_learn(seastar::rpc::client_info const&, seastar::rpc::opt_time_point, service::paxos::proposal, utils::small_vector<gms::inet_address, 3ul>, gms::inet_address, unsigned int, unsigned long, std::optional<tracing::trace_info>)::{lambda(seastar::shared_ptr<service::storage_proxy>&, tracing::trace_state_ptr, seastar::lw_shared_ptr<schema const>, service::paxos::proposal const&, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >, service::fencing_token)#1}&&, service::storage_proxy::remote::handle_paxos_learn(seastar::rpc::client_info const&, seastar::rpc::opt_time_point, service::paxos::proposal, utils::small_vector<gms::inet_address, 3ul>, gms::inet_address, unsigned int, unsigned long, std::optional<tracing::trace_info>)::{lambda(seastar::shared_ptr<service::storage_proxy>&, netw::msg_addr, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >, service::paxos::proposal const&, gms::inet_address, unsigned int, unsigned long, std::optional<tracing::trace_info> const&, service::fencing_token)#1}&&) at ./service/storage_proxy.cc:549
seastar::coroutine::all<seastar::future<void>, seastar::future<void> >::intermediate_task<0ul>::run_and_dispose() at ././seastar/include/seastar/coroutine/all.hh:138
```

```
0x608000c293a8 is located 8 bytes inside of 96-byte region [0x608000c293a0,0x608000c29400)
freed by thread T0 here:
    #0 0xd268b21  (/jenkins/workspace/scylla-master/dtest-debug/scylla/.ccm/scylla-repository/4e18f8b45381c6b23cf271cc686194d87bf212a1/libexec/scylla+0xd268b21) (BuildId: d5c28c468be96f95c71317049ad35cd1814a67dc)
    #1 0x10db13fe  (/jenkins/workspace/scylla-master/dtest-debug/scylla/.ccm/scylla-repository/4e18f8b45381c6b23cf271cc686194d87bf212a1/libexec/scylla+0x10db13fe) (BuildId: d5c28c468be96f95c71317049ad35cd1814a67dc)
    #2 0x7f73352f1558  (/jenkins/workspace/scylla-master/dtest-debug/scylla/.ccm/scylla-repository/4e18f8b45381c6b23cf271cc686194d87bf212a1/libreloc/libseastar.so+0x46f1558) (BuildId: 95698310d1473fda2254677214f6407401d5374d)
    #3 0x7f73352f55f0  (/jenkins/workspace/scylla-master/dtest-debug/scylla/.ccm/scylla-repository/4e18f8b45381c6b23cf271cc686194d87bf212a1/libreloc/libseastar.so+0x46f55f0) (BuildId: 95698310d1473fda2254677214f6407401d5374d)
    #4 0x7f73352f3273  (/jenkins/workspace/scylla-master/dtest-debug/scylla/.ccm/scylla-repository/4e18f8b45381c6b23cf271cc686194d87bf212a1/libreloc/libseastar.so+0x46f3273) (BuildId: 95698310d1473fda2254677214f6407401d5374d)
    #5 0x7f7334e2caba  (/jenkins/workspace/scylla-master/dtest-debug/scylla/.ccm/scylla-repository/4e18f8b45381c6b23cf271cc686194d87bf212a1/libreloc/libseastar.so+0x422caba) (BuildId: 95698310d1473fda2254677214f6407401d5374d)
    #6 0x7f7334e2a05a  (/jenkins/workspace/scylla-master/dtest-debug/scylla/.ccm/scylla-repository/4e18f8b45381c6b23cf271cc686194d87bf212a1/libreloc/libseastar.so+0x422a05a) (BuildId: 95698310d1473fda2254677214f6407401d5374d)
    #7 0xd270d49  (/jenkins/workspace/scylla-master/dtest-debug/scylla/.ccm/scylla-repository/4e18f8b45381c6b23cf271cc686194d87bf212a1/libexec/scylla+0xd270d49) (BuildId: d5c28c468be96f95c71317049ad35cd1814a67dc)
    #8 0xd26dc2f  (/jenkins/workspace/scylla-master/dtest-debug/scylla/.ccm/scylla-repository/4e18f8b45381c6b23cf271cc686194d87bf212a1/libexec/scylla+0xd26dc2f) (BuildId: d5c28c468be96f95c71317049ad35cd1814a67dc)
    #9 0x7f732ea49b89  (/jenkins/workspace/scylla-master/dtest-debug/scylla/.ccm/scylla-repository/4e18f8b45381c6b23cf271cc686194d87bf212a1/libreloc/libc.so.6+0x27b89) (BuildId: 7026fe8c129a523e07856d7c96306663ceab6e24)
    #10 0x7f732ea49c4a  (/jenkins/workspace/scylla-master/dtest-debug/scylla/.ccm/scylla-repository/4e18f8b45381c6b23cf271cc686194d87bf212a1/libreloc/libc.so.6+0x27c4a) (BuildId: 7026fe8c129a523e07856d7c96306663ceab6e24)
    #11 0xd191c64  (/jenkins/workspace/scylla-master/dtest-debug/scylla/.ccm/scylla-repository/4e18f8b45381c6b23cf271cc686194d87bf212a1/libexec/scylla+0xd191c64) (BuildId: d5c28c468be96f95c71317049ad35cd1814a67dc)
```

Decoded:
```
operator delete(void*) at ??:?
std::default_delete<service::storage_proxy::remote>::operator()(service::storage_proxy::remote*) const at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/unique_ptr.h:99
 (inlined by) std::__uniq_ptr_impl<service::storage_proxy::remote, std::default_delete<service::storage_proxy::remote> >::reset(service::storage_proxy::remote*) at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/unique_ptr.h:211
 (inlined by) std::unique_ptr<service::storage_proxy::remote, std::default_delete<service::storage_proxy::remote> >::reset(service::storage_proxy::remote*) at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/unique_ptr.h:509
 (inlined by) std::unique_ptr<service::storage_proxy::remote, std::default_delete<service::storage_proxy::remote> >::operator=(decltype(nullptr)) at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/unique_ptr.h:442
 (inlined by) service::storage_proxy::stop_remote() at ./service/storage_proxy.cc:6332
```

```
previously allocated by thread T0 here:
==1475==WARNING: Can't read from symbolizer at fd 31
==1475==WARNING: Can't read from symbolizer at fd 31
    #0 0xd2682a1  (/jenkins/workspace/scylla-master/dtest-debug/scylla/.ccm/scylla-repository/4e18f8b45381c6b23cf271cc686194d87bf212a1/libexec/scylla+0xd2682a1) (BuildId: d5c28c468be96f95c71317049ad35cd1814a67dc)
    #1 0x10ca5243  (/jenkins/workspace/scylla-master/dtest-debug/scylla/.ccm/scylla-repository/4e18f8b45381c6b23cf271cc686194d87bf212a1/libexec/scylla+0x10ca5243) (BuildId: d5c28c468be96f95c71317049ad35cd1814a67dc)
```

Decoded:
```
operator new(unsigned long) at ??:?
std::__detail::_MakeUniq<service::storage_proxy::remote>::__single_object std::make_unique<service::storage_proxy::remote, service::storage_proxy&, netw::messaging_service&, gms::gossiper&, service::migration_manager&, seastar::sharded<db::system_keyspace>&>(service::storage_proxy&, netw::messaging_service&, gms::gossiper&, service::migration_manager&, seastar::sharded<db::system_keyspace>&) at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/unique_ptr.h:1070
 (inlined by) service::storage_proxy::start_remote(netw::messaging_service&, gms::gossiper&, service::migration_manager&, seastar::sharded<db::system_keyspace>&) at ./service/storage_proxy.cc:6327
```

**Comment by bhalevy** [comment]
It looks like handle_write (and possible other callbacks) needs to enter a gate
that should be closed in `remote::stop` so that `stop_remote` would wait on async fibers before destroying `_remote`.

**Comment by bhalevy** [comment]
@kostja / @kbr-scylla who should won this issue?

**Comment by kostja** [comment]
* Benny Halevy ***@***.***> [24/01/14 19:54]:
> @kostja / @kbr-scylla who should own this issue?

It is us.

-- 
Konstantin Osipov, Moscow, Russia

**Comment by mykaul** [comment]
@kbr-scylla - is this on track to 6.0?

**Comment by kbr-scylla** [comment]
It looks like low priority. I wouldn't make it a blocker for 6.0. We're focusing on delivering raft-topology.
